File: src/memory.c
#include <time.h>
#include <stdbool.h>
#include <stdio.h>

#include "memory.h"
#include "list.h"
#include "macros.h"

#define PHYSICAL 1
#define VIRTUAL -1
#define NOT_IN_MEM 0 

struct page {
    time_t      counter;
    int         location;
    int         index;
};

static struct page page_table[TOTAL_MEM] = {{0}};
static struct page *phys_mem[PHYSICAL_MEM] = {NULL};
static struct page *virt_mem[VIRTUAL_MEM] = {NULL};

static int mem_remaining = TOTAL_MEM;

bool mem_check(int size) {
    return size < mem_remaining;
}

list mem_allocate(int size) {
    DEBUG_PRINT("[Memory] Entering memory alloc");
    if (size > mem_remaining)
        return NULL;

    list page_list = l_init();

    DEBUG_PRINT("[Memory] List initialized");
    for (int i = 0; i < TOTAL_MEM; i++) {
        if (!page_table[i].location) {
            DEBUG_PRINT("[Memory] Found page");
            page_table[i].index = -1;
            for (int j = 0; j < PHYSICAL_MEM; j++) {
                if (!phys_mem[j]) {
                    DEBUG_PRINT("[Memory] Allocating in physical memory");
                    phys_mem[j] = &page_table[i];
                    page_table[i].index = j;
                    page_table[i].location = PHYSICAL;
                    break;
                }
            }
            if (page_table[i].index == -1) {
                for (int j = 0; j < VIRTUAL_MEM; j++) {
                    if (!virt_mem[j]) {
                        DEBUG_PRINT("[Memory] Allocating in virtual memory");
                        virt_mem[j] = &page_table[i];
                        page_table[i].index = j;
                        page_table[i].location = VIRTUAL;
                        break;
                    }
                }
            }
            page_table[i].counter = time(0);
            l_add(page_list, &page_table[i]);
            mem_remaining--;
            if (l_size(page_list) == size)
                break;
        }
    }

    return page_list;
}

void mem_deallocate(list page_list) {
    //DEBUG_PRINT("[Memory] Deallocating memory");
    for (int i = 0; i < l_size(page_list); i++) {
        struct page *p = l_get(page_list, i);
        if (p->location == PHYSICAL) {
            phys_mem[p->index] == NULL;
        }
        else {
            virt_mem[p->index] == NULL;
        }

        p->location = NOT_IN_MEM;
        mem_remaining++;
    }
}

static struct page *findLeastUsed() {
    int lu_ind;
    time_t min;
    bool first_time = true;

    for (int i = 0; i < PHYSICAL_MEM; i++) {
        if (phys_mem[i]) {
            if (first_time) {
                min = phys_mem[i]->counter;
                lu_ind = i;
                first_time = false;
            }
            else {
                if (phys_mem[i]->counter < min) {
                    min = phys_mem[i]->counter;
                    lu_ind = i;
                }
            }
        }
    }

    return phys_mem[lu_ind];
}

void mem_access(list page_list) {
    for (int i = 0; i < l_size(page_list); i++) {
        struct page *p = l_get(page_list, i);

        if (p->location == VIRTUAL) {
            struct page *swap_p = findLeastUsed();
            DEBUG_PRINT("[Memory] Swapping page into slot %d", swap_p->index);
            phys_mem[swap_p->index] = p;
            virt_mem[p->index] = swap_p;

            int temp = p->index;
            p->index = swap_p->index;
            swap_p->index = temp;
        }

        p->counter = time(0);
    }
}

File: src/memory.py
from threading import Lock
from time import time

import src.exceptions as ex

from src.globals import DebugPrint

class Page:
    def __init__(self):
        self.counter = 0
        self.location = None

class Memory:
    PHYSICAL_MEM = 4096
    VIRTUAL_MEM = 8192
    CACHE_MEM = 40
    TOTAL_MEM = PHYSICAL_MEM + VIRTUAL_MEM + CACHE_MEM

    def __init__(self):
        self.page_table = [Page()] * self.TOTAL_MEM
        self.cache = [None] * self.CACHE_MEM
        self.phys = [None] * self.PHYSICAL_MEM
        self.virt = [None] * self.VIRTUAL_MEM
        self.lock = Lock()

    def remaining(self):
        return self.phys.count(None) + self.virt.count(None)
    
    def full(self):
        return False if None in self.phys and None in self.virt else True

    def allocate(self, size):
        with self.lock:
            DebugPrint("[Memory] Entering memory alloc")
            if size > self.remaining():
                raise ex.MemoryAllocationError

            page_list = list()
            for page in self.page_table:
                if page.location is None:
                    if None not in self.phys:
                        if None not in self.virt:
                            table = self.cache
                        else:
                            table = self.virt
                    else:
                        table = self.phys
                    index = table.index(None)
                    table[index] = page
                    page.location = table
                    page_list.append(page)
                if len(page_list) == size:
                    break

        return page_list

    def deallocate(self, page_list):
        with self.lock:
            for page in page_list:
                index = page.location.index(page)
                page.location[index] = None
                page.location = None

    def findLeastUsed(self, table):
        min = table[0]
        for page in table:
            if page.counter < min.counter:
                min = page

        return page

    def access(self, page_list):
        with self.lock:
            start_page = page_list[0]
            start_loc = start_page.location
            if start_loc is not self.cache:
                swap_p = self.findLeastUsed(self.cache)
                self.cache[self.cache.index(swap_p)] = start_page
                start_loc[start_loc.index(start_page)] = swap_p
                swap_p.location = start_loc
                start_page.location = self.cache

            for page in page_list:
                page.counter = time()
                if page.location is self.virt:
                    swap_p = self.findLeastUsed(self.phys)
                    self.phys[self.phys.index(swap_p)] = page
                    self.virt[self.virt.index(page)] = swap_p
                    swap_p.location = self.virt
                    page.location = self.phys

memory = Memory()


